polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs

```diff
@@ -0,0 +1,1047 @@
+		&self.base_constraints
+	}
+}
+
+/// A hypothetical candidate, which may or may not exist in
+/// the fragment chain already.
+pub(crate) enum HypotheticalCandidate<'a> {
+	Complete {
+		receipt: Cow<'a, CommittedCandidateReceipt>,
+		persisted_validation_data: Cow<'a, PersistedValidationData>,
+	},
+	Incomplete {
+		relay_parent: Hash,
+		parent_head_data_hash: Hash,
+	},
+}
+
+impl<'a> HypotheticalCandidate<'a> {
+	fn parent_head_data_hash(&self) -> Hash {
+		match *self {
+			HypotheticalCandidate::Complete { ref persisted_validation_data, .. } =>
+				persisted_validation_data.as_ref().parent_head.hash(),
+			HypotheticalCandidate::Incomplete { ref parent_head_data_hash, .. } =>
+				*parent_head_data_hash,
+		}
+	}
+
+	fn output_head_data_hash(&self) -> Option<Hash> {
+		match *self {
+			HypotheticalCandidate::Complete { ref receipt, .. } =>
+				Some(receipt.descriptor.para_head),
+			HypotheticalCandidate::Incomplete { .. } => None,
+		}
+	}
+
+	fn relay_parent(&self) -> Hash {
+		match *self {
+			HypotheticalCandidate::Complete { ref receipt, .. } =>
+				receipt.descriptor().relay_parent,
+			HypotheticalCandidate::Incomplete { ref relay_parent, .. } => *relay_parent,
+		}
+	}
+}
+
+impl<'a> From<&'a messages::HypotheticalCandidate> for HypotheticalCandidate<'a> {
+	fn from(value: &'a messages::HypotheticalCandidate) -> Self {
+		match value {
+			messages::HypotheticalCandidate::Complete {
+				receipt,
+				persisted_validation_data,
+				..
+			} => Self::Complete {
+				receipt: Cow::Borrowed(receipt),
+				persisted_validation_data: Cow::Borrowed(persisted_validation_data),
+			},
+			messages::HypotheticalCandidate::Incomplete {
+				parent_head_data_hash,
+				candidate_relay_parent,
+				..
+			} => Self::Incomplete {
+				relay_parent: *candidate_relay_parent,
+				parent_head_data_hash: *parent_head_data_hash,
+			},
+		}
+	}
+}
+
+pub struct FragmentNode {
+	fragment: Fragment<'static>,
+	pub candidate_hash: CandidateHash,
+	cumulative_modifications: ConstraintModifications,
+}
+
+impl FragmentNode {
+	fn relay_parent(&self) -> Hash {
+		self.fragment.relay_parent().hash
+	}
+}
+
+/// This is a chain of candidates based on some underlying storage of candidates and a scope.
+///
+/// All nodes in the chain must be either pending availability or within the scope. Within the scope
+/// means it's built off of the relay-parent or an ancestor.
+pub(crate) struct FragmentChain {
+	scope: Scope,
+
+	pub chain: Vec<FragmentNode>,
+
+	candidates: HashSet<CandidateHash>,
+
+	// Index from head data hash to candidate hashes with that head data as a parent.
+	by_parent_head: HashMap<Hash, CandidateHash>,
+	// Index from head data hash to candidate hashes outputting that head data.
+	by_output_head: HashMap<Hash, CandidateHash>,
+}
+
+impl FragmentChain {
+	/// Create a new [`FragmentChain`] with given scope and populated from the storage.
+	pub fn populate(scope: Scope, storage: &CandidateStorage) -> Self {
```

Comment:
Not sure this is a good name for constructor. Why not just `new` ?

Reply:
I find `populate` more expressive in this case. Also, this name wasn't introduced here